README.md
# Cookie-Sessions

Secure cookie-based session middleware for Express.

Session data is stored on the request object in the 'session' property:
```js
  var app = require('express');
  var cookieParser = require('cookie-parser');
  var cookieSessions = require('cookie-sessions');

  app.use(
    cookieSessions({
      name: 'session_data',
      secret: process.env.SECRET
    })
  );
```

The [cookie-parser](https://www.npmjs.com/package/cookie-parser) middleware
MUST also used.

The session data can be any JSON object. It's timestamped, encrypted and
authenticated automatically. The authenticated encryption uses `aes-256-gcm`
offered by the node `crypto` library. The httpOnly and secure cookie flags are
set by default.

The main function accepts a number of options:

| Option        | Required | Description                                                                                                             | Default  |
|---------------|----------|-------------------------------------------------------------------------------------------------------------------------|----------|
| secret        | Yes      | The secret to encrypt the session data.                                                                                 |          |
| timeout       | Yes      | The amount of time in milliseconds before the cookie expires.                                                           | 24 hours |
| name          | Yes      | The cookie name in which to store the session data.                                                                     | `\_node` |
| path          | Yes      | The path to use for the cookie.                                                                                         | `/`      |
| domain        | No       | Define a specific domain/subdomain scope for the cookie.                                                                |          |
| autoRenew     | No       | Boolean: if true, a new cookie will be set in each response with an updated expiration Date.now() + timeout             |   true   |
| httpOnly      | No       | Boolean: if true, the httpOnly cookie flag will be set.                                                                 |   true   |
| secure        | No       | Boolean: if true, the secure cookie flag will be set.                                                                   |   true   |
| sameSite      | No       | If set to "lax" or "strict", the sameSite cookie flag with the corresponding mode will be set.                          |          |
| sessionCookie | No       | Boolean: if true, it's considered a session cookie and no "expires" is set.                                             |          |


## Why store session data in cookies?

* Its fast, you don't need to hit the filesystem or a database to look up
  session data
* It scales easily. You don't need to worry about sticky-sessions when
  load-balancing across multiple nodes.
* No server-side persistence requirements

## Caveats

* You can only store 4k of data in a cookie
* Higher-bandwidth requirements, since the cookie is sent to the server with
  every request.

__In summary:__ don't use cookie storage if you keep a lot of data in your
sessions!

## Migrating to version 1.0.0

* Any cookie created with 0.0.2 version will be invalidated.
* The `options` object has two naming changes:
  * `name` instead of `session_key`
  * `sessionCookie` instead of `session_cookie`
* The following exported functions have been removed:
  * readSession
  * readCookies
  * checkLength
  * headersToArray
  * hmac\_signature


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## v1.0.1 - 2019-07-17

### Misc

- Minor update in lodash to mitigate a snyk reported vulnerability
- Fixes bug in tests
- Minor updates in README

## v1.0.0 - 2018-10-30

### Adds

- Adds option for the samesite cookie flag
- Adds autoRenew option

### Changes

- Updates cryptographic algorithms. It's now using AES 256 in GCM mode

### Removes

- Removes the following exported functions:
  - readSession
  - readCookies
  - checkLength
  - headersToArray
  - hmac_signature

### Misc

- Code refactor


test.js
#!/usr/local/bin/node

try {
    var testrunner = require('./deps/nodeunit').testrunner;
}
catch(e) {
    var sys = require('sys');
    sys.puts("Cannot find nodeunit module.");
    sys.puts("You can download submodules for this project by doing:");
    sys.puts("");
    sys.puts("    git submodule init");
    sys.puts("    git submodule update");
    sys.puts("");
    process.exit();
}

process.chdir(__dirname);
testrunner.run(['test']);


test/cookie-sessions.test.js
var sessions = require('../lib/cookie-sessions');
var assert  = require('assert')
var sinon = require('sinon')
var crypto = require('crypto')
var request = require('supertest');
var express = require('express');
var cookieParser = require('cookie-parser')

describe('cookie sessions tests', function() {
  var secret = '8c06bdc84bf095d76b18c1e3a485dfe6'
  var message = 'the answer to the ultimate question of life, the universe, and everything'
  var encr = '017cb25516cf63f676e680abf99e2797da7f0765b3f19e53adbdadfaf2cd8f9a9378a35b1e52fb00e2a5aa577eb277d98096dbb75afd9dd57b0498e54ead8fc66ed980bf1e60a1b2b8'
  var iv = '55df03d8dd90145b7f23211613d46b2d'
  var authTag = 'c9e0fe9af2412ce941d18e7eae667402'
  var user_data = {
    id: 123,
    username: 'foobar',
    email: 'foobar@gmail.com',
    role: 'user'
  }

  beforeEach(function () {
    this.sandbox = sinon.createSandbox();
  });

  afterEach(function () {
    this.sandbox.restore();
  });

  describe('encrypt', function() {
    it('should sucessfully return encrypted data', function() {
      enc = sessions.encrypt(secret, message)
      assert.ok(enc.ciphertext)
      assert.ok(enc.iv)
      assert.ok(enc.authTag)
    })
  })

  describe('decrypt', function() {
    it('should sucessfully return decrypted data', function() {
      plaintext = sessions.decrypt(secret, encr, iv, authTag)
      assert.equal(message, plaintext)
    })
  })

  describe('serialize', function() {
    it('should return a string', function() {
      var result = sessions.serialize(secret, user_data, 1540678757323,  false)
      var arr = result.split('$')
      var cipher = arr[0]
      var iv = arr[1]
      var authTag = arr[2]
      assert.ok(cipher)
      assert.ok(iv)
      assert.ok(authTag)
    })

    it('should return error on long data', function() {
      var  long = '';
      for(var i = 0; i < 1500; i++){
          long += 'a';
      };
      this.sandbox.stub(sessions, 'encrypt').returns({ciphertext: long, iv: long, authTag: long});
      try {
        var cookiestr = sessions.serialize(secret, user_data)
      }
      catch(err) {
        assert.equal(err.message, 'Data too long to store in a cookie')
      };
    })
  })

  describe('deserialize', function() {
    it('should return user data in JSON if cookie has not expired', function() {
      var cookiestr = sessions.serialize(secret, user_data, Date.now(), false)
      var jsonData = sessions.deserialize(secret, cookiestr, 1000 * 60 * 60 * 12)
      assert.deepEqual(jsonData.data, user_data)
    })

    it('should return validation error if cookie has expired', function() {
      var str = '3d11d4581d2359247a4183fc3c84aceaa5835d38075117b10c1f8bcff4a63f0d3740f6ff30f60942b9851af38fd8f1d5bcf6b2376c6803f001a844dc71b5d50276cbb735f501c47d320ef6c6d07cd2047b7a6b26521c23687718ebf248bd9b$69d2b05d79b47c3e5a94b0f32f2fbe3b$ed4c06a57df5a2b646f7e5f82bffb17f'
      try {
        var jsonData = sessions.deserialize(secret, str, 1000 * 60 * 60 * 10)
      }
      catch(err) {
        assert.ok(err)
        assert.equal(err.message, 'Cookie expired')
      }
    })
  })

  describe('onInit', function() {
    it('throw error if no secret set in server settings', function() {
      try {
        sessions({})
      } catch(err) {
        assert.ok(err)
        assert.equal(err.message, 'No secret set in cookie-session settings')
      }
    });
    it('throw error for invalid path', function() {
      try {
        sessions({secret: secret, path: 'foo/bar'})
      } catch(err) {
        assert.ok(err)
        assert.equal(err.message, 'Invalid cookie path, must start with "/"')
      }
    });
    it('should do nothing for a request not under the specified path', (done) => {
      const session = sessions({path: '/foo/bar', secret: secret })
      req = { path: '/foo/baz' }
      res = {}
      next = function() {
        assert.ok(!req.session)
        done()
      }
      session(req, res, next)
    });
    it('throw error for invalid value in sameSite option', function() {
      try {
        sessions({secret: secret, sameSite: true})
      } catch(err) {
        assert.ok(err)
        assert.equal(err.message, 'Possible values for sameSite option: "lax" or "strict"')
      }
    });
  })

  describe('the middleware', function() {
    beforeEach(function () {
      this.app = express();
    });

    describe('when session has already data', function() {
      beforeEach(function() {
        this.cookieTime = Date.now()
        this.sandbox.stub(sessions, 'deserialize').returns({data: {username: 'foobar'}, time: this.cookieTime});
      });

      describe('and new data are added to req.session', function() {
        describe('and autoRenew is set to true', function() {
          it('should set a new cookie with the combined data and with new timestamp', function () {
            this.app.use(cookieParser())
            this.app.use(sessions({secret: secret, autoRenew: true}))
            this.app.get('/', (req, res) => {
              req.session.koko = 'bar'
              res.send('Hello World!')
            })
            return request(this.app)
            .get('/')
            .then(res => {
              this.sandbox.restore();
              assert(res.headers['set-cookie'])
              _node = res.headers['set-cookie'][0].split(';')[0].split('=')[1]
              var jsonData = sessions.deserialize(secret, decodeURIComponent(_node), 1000 * 60 * 60 * 24)
              assert.notEqual(jsonData.time, this.cookieTime)
              assert.deepEqual(jsonData.data, { username: 'foobar', koko: 'bar' })
            })
          });
        });
        describe('and autoRenew is set to false', function() {
          it('should set a new cookie with the combined data and the initial timestamp', function () {
            this.app.use(cookieParser())
            this.app.use(sessions({secret: secret, autoRenew: false}))
            this.app.get('/', (req, res) => {
              req.session.koko = 'bar'
              res.send('Hello World!')
            })
            return request(this.app)
            .get('/')
            .then(res => {
              this.sandbox.restore();
              assert(res.headers['set-cookie'])
              var _node = res.headers['set-cookie'][0].split(';')[0].split('=')[1]
              var jsonData = sessions.deserialize(secret, decodeURIComponent(_node), 1000 * 60 * 60 * 24)
              assert.equal(jsonData.time, this.cookieTime)
              assert.deepEqual(jsonData.data, { username: 'foobar', koko: 'bar' })
            })
          });
        });
      });
      describe('and no data are added to req.session', function() {
        describe('and autoRenew is set to true', function() {
          it('should set a new cookie with new timestamp', function () {
            this.app.use(cookieParser())
            this.app.use(sessions({secret: secret, autoRenew: true}))
            this.app.get('/', (req, res) => {
              res.send('Hello World!')
            })
            return request(this.app)
            .get('/')
            .then(res => {
              this.sandbox.restore();
              assert(res.headers['set-cookie'])
              var _node = res.headers['set-cookie'][0].split(';')[0].split('=')[1]
              var jsonData = sessions.deserialize(secret, decodeURIComponent(_node), 1000 * 60 * 60 * 24)
              assert.notEqual(jsonData.time, this.cookieTime)
              assert.deepEqual(jsonData.data, { username: 'foobar'})
            })
          });
        });
        describe('and autoRenew is set to false', function() {
          it('should not set a new cookie', function () {
            this.app.use(cookieParser())
            this.app.use(sessions({secret: secret, autoRenew: false}))
            this.app.get('/', (req, res) => {
              res.send('Hello World!')
            })
            return request(this.app)
            .get('/')
            .then(res => {
              this.sandbox.restore();
              assert(!res.headers['set-cookie'])
            })
          });
        });
      });
      describe('data get deleted from req.session', function() {
        it('should set a new cookie with empty data', function () {
          this.app.use(cookieParser())
          this.app.use(sessions({secret: secret}))
          this.app.get('/', (req, res) => {
            req.session = {}
            res.send('Hello World!')
          })
          this.sandbox.restore();
          var cookiestr = sessions.serialize(secret, user_data, Date.now(),  false)
          return request(this.app)
          .get('/')
          .set('Cookie', '_node=' + encodeURIComponent(cookiestr))
          .then(res => {
            assert(res.headers['set-cookie'])
            var _node = res.headers['set-cookie'][0].split(';')[0].split('=')[1]
            var jsonData = sessions.deserialize(secret, decodeURIComponent(_node), 1000 * 60 * 60 * 24)
            assert.deepEqual(jsonData.data, {})
          })
        });
      });
    });
    describe('when session is empty', function() {
      describe('and new data are added to req.session', function() {
        it('should set a new cookie with the new data and with a timestamp', function () {
          this.sandbox.stub(sessions, 'deserialize').returns({data: {}});
          this.app.use(cookieParser())
          this.app.use(sessions({secret: secret}))
          this.app.get('/', (req, res) => {
            req.session.koko = 'bar'
            res.send('Hello World!')
          })
          return request(this.app)
          .get('/')
          .then(res => {
            this.sandbox.restore()
            assert(res.headers['set-cookie'])
            var _node = res.headers['set-cookie'][0].split(';')[0].split('=')[1]
            var jsonData = sessions.deserialize(secret, decodeURIComponent(_node), 1000 * 60 * 60 * 24)
            assert(jsonData.time)
            assert.deepEqual(jsonData.data, { koko: 'bar' })
          })
        });
      });
    });
  })
})


lib/cookie-sessions.js
var _ = require('lodash')
var crypto = require('crypto')

const MAX_LENGTH = 4096;

// don't set a default cookie secret, must be explicitly defined
const DEFAULT_SETTINGS = {
  name: '_node',
  timeout: 1000 * 60 * 60 * 24, // 24 hours
  path: '/',
  autoRenew: true,
  httpOnly: true,
  secure: true
};

var exports = module.exports = function(settings) {
  var opts = Object.assign({}, DEFAULT_SETTINGS, settings)

  if (!opts.secret) {
    throw new Error('No secret set in cookie-session settings');
  }

  if (typeof opts.path !== 'string' || opts.path.indexOf('/') !== 0) {
    throw new Error('Invalid cookie path, must start with "/"');
  }

  if (opts.sameSite && (opts.sameSite !== 'lax' || opts.sameSite !== 'strict')) {
    throw new Error('Possible values for sameSite option: "lax" or "strict"');
  }

  return function(req, res, next) {

    // if the request is not under the specified path, do nothing.
    if (req.path.indexOf(opts.path) !== 0) {
      next();
      return;
    }
    req.session = {}

    // Read session data from a request and store it in req.session
    var session = {
      data: {}
    }
    try {
      session = exports.deserialize(opts.secret, req.cookies[opts.name], opts.timeout);
      req.session = session.data
    } catch (err) {
    }
    var oldSession = _.cloneDeep(req.session)

    var options = _.pick(opts, [ 'path', 'domain', 'httpOnly', 'secure', 'sameSite' ])

    var _writeHead = res.writeHead;
    res.writeHead = function() {
      // only set cookie if autoRenew is on or session data changed
      if (opts.autoRenew || !_.isEqual(oldSession, req.session)) {
        /*
         * If the session is empty, no cookie is sent (and the cookie is
         * cleared if it is already set). If a session exists, we add a
         * Set-Cookie header to all responses with the session data and the
         * current timestamp. The cookie needs to be set on every response so
         * that the timestamp is up to date, and the session does not expire
         * unless the user is inactive.
         */
        if (_.isEqual(req.session, {})) {
          if (req.cookies[opts.name]) {
            res.clearCookie(opts.name, options)
          }
        } else {
          if ((!opts.sessionCookie || !opts.autoRenew) && opts.timeout) {
            options.expires = new Date(Date.now() + opts.timeout);
          }
          try {
            cookiestr = exports.serialize(opts.secret, req.session, session.time, opts.autoRenew)
          } catch(err) {
            cookiestr = ''
          }
          res.cookie(opts.name, cookiestr, options)
        }
      }

      return _writeHead.apply(res, arguments);
    }
    next()
  };
};

exports.split = function(str){
  var arr = str.split('$')
  return {
    ciphertext: arr[0],
    iv: arr[1],
    authTag: arr[2]
  };
};

exports.merge = function(obj) {
  return [ obj.ciphertext, obj.iv, obj.authTag ].join('$')
};

exports.valid = function(timestamp, timeout) {
  return timestamp + timeout > Date.now();
}

exports.encrypt = function(secret, message) {
  try {
    var iv = crypto.randomBytes(16);
    var cipher = crypto.createCipheriv('aes-256-gcm', secret, iv);
    var ciphertext = cipher.update(message, 'utf8', 'hex');
    ciphertext += cipher.final('hex');
    return {
      ciphertext: ciphertext,
      iv: iv.toString('hex'),
      authTag: cipher.getAuthTag().toString('hex')
    };
  } catch(err) {
    throw new Error('Failed to encrypt cookie')
  }
};

exports.decrypt = function(secret, ciphertext, iv, authTag) {
  try {
    var decipher = crypto.createDecipheriv('aes-256-gcm', secret, Buffer.from(iv, 'hex'));
    decipher.setAuthTag(Buffer.from(authTag, 'hex'));
    var plaintext = decipher.update(ciphertext, 'hex', 'utf8');
    plaintext += decipher.final('utf8');
    return plaintext;
  } catch(err) {
    throw new Error('Failed to decrypt and authenticate cookie')
  }
};

// Turns a JSON-compatibile object literal into a secure cookie string
exports.serialize = function(secret, data, sessionTime, autoRenew) {
  if (autoRenew || typeof sessionTime === 'undefined') {
    sessionTime = Date.now()
  }
  var cookieData = {
    d: data,
    t: sessionTime
  }
  var dataStr = JSON.stringify(cookieData);
  var enc = exports.encrypt(secret, dataStr)
  var result = exports.merge(enc)
  if (result.length > MAX_LENGTH) {
    throw Error('Data too long to store in a cookie') ;
  }
  return result
}

// Parses a secure cookie string, returning the object stored within it.
exports.deserialize = function(secret, str, timeout) {
  if (!str) {
    return {
      data: {}
    }
  }
  var splitted = exports.split(str)
  cookieData = exports.decrypt(secret, splitted.ciphertext, splitted.iv, splitted.authTag)
  var jsonData = JSON.parse(cookieData)

  if (!exports.valid(jsonData.t, timeout)) {
    throw new Error('Cookie expired')
  }
  return {
    data: jsonData.d,
    time: jsonData.t
  }
};
